metastore/src/java/org/apache/hadoop/hive/metastore/hbase/StatsCache.java

```diff
@@ -0,0 +1,324 @@
+   * @param colName of column to get stats for
+   * @return stats object for this column, or null if none cached
+   */
+  StatsInfo getTableStatistics(String dbName, String tableName, String colName) {
+    return getStatistics(new Key(dbName, tableName, colName), tableHits, tableMisses);
+  }
+
+  /**
+   * Get partition level statistics
+   * @param dbName name of database table is in
+   * @param tableName name of table
+   * @param partName name of this partition
+   * @param colName of column to get stats for
+   * @return stats object for this column, or null if none cached
+   */
+  StatsInfo getPartitionStatistics(String dbName, String tableName,
+                                             String partName, String colName) {
+    return getStatistics(new Key(dbName, tableName, partName, colName), partHits, partMisses);
+  }
+
+  String[] dumpMetrics() {
+    String[] strs = new String[counters.size()];
+    for (int i = 0; i < strs.length; i++) {
+      strs[i] = counters.get(i).dump();
+    }
+    return strs;
+  }
+
+  private StatsInfo getStatistics(Key key, Counter hits, Counter misses) {
+    StatsInfo s = null;
+    lock.readLock().lock();
+    try {
+      s = cache.get(key);
+    } finally {
+      lock.readLock().unlock();
+    }
+    if (s == null) {
+      misses.incr();
+      return null;
+    } else if (tooLate(s)) {
+      remove(key);
+      misses.incr();
+      return null;
+    } else {
+      s.lastTouched = System.currentTimeMillis();
+      hits.incr();
+      return s;
+    }
+  }
+
+  private void remove(Key key) {
+    lock.writeLock().lock();
+    try {
+      // It's possible that multiple callers will call remove for a given key, so don't complain
+      // if the indicated key is already gone.
+      cache.remove(key);
+    } finally {
+      lock.writeLock().unlock();
+    }
+  }
+
+  private void clean() {
+    // This spawns a separate thread to walk through the cache and clean.
+    synchronized (this) {
+      if (cleaning) return;
+      cleaning = true;
+      cleans.incr();
+    }
+
+      Thread cleaner = new Thread() {
+        @Override
+        public void run() {
+          try {
+            // Get the read lock and then make a copy of the map.  This is so we can work through it
+            // without having concurrent modification exceptions.  Then walk through and remove things
+            // one at a time.
+            List<Map.Entry<Key, StatsInfo>> entries = null;
+            lock.readLock().lock();
+            try {
+              entries = new ArrayList<Map.Entry<Key, StatsInfo>>(cache.entrySet());
+            } finally {
+              lock.readLock().unlock();
+            }
+            for (Map.Entry<Key, StatsInfo> entry : entries) {
+              if (tooLate(entry.getValue())) {
+                remove(entry.getKey());
+              }
+              // We want to make sure this runs at a low priority in the background
+              Thread.yield();
```

Comment:
This Thread.yield() is within a for-loop, that might consume some cpu cycles as well. I wonder if just setting the priority is sufficient in most cases. Since there is going to be only one cleaner thread for this cache, I wonder if it is OK to not call yield at all. The javadocs for it recommend its use only if the profiling results suggest its need.